nemo/collections/nlp/models/language_modeling/megatron_lm_adapter_encoder_decoder_model.py

```diff
@@ -0,0 +1,166 @@
+        for _, module in self.enc_dec_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    module.add_adapter(name=adapter_key, cfg=self.adapter_cfg)
+        logging.info(f'After adding adapters:\n{self.summarize()}')
+
+        # load adapters weights if provided
+        if hasattr(self.cfg, 'adapters_file'):
+            self.loading_base_model = False
+            with tempfile.TemporaryDirectory() as tmpdir:
+                adapters_state_dict = self.extract_state_dict_from(
+                    self.cfg.adapters_file, tmpdir, save_restore_connector=NLPSaveRestoreConnector()
+                )
+                self._load_adapters_weights(adapters_state_dict)
+                logging.info(f'Adapters weights loaded successfully from {self.cfg.adapters_file}')
+            self.loading_base_model = True
+
+    def _validate_adapters_cfg(self, cfg):
+        assert cfg.type in ['parallel_adapter', 'linear_adapter']
+        assert hasattr(cfg, 'adapter_dim')
+        assert cfg.norm_position in ['pre', 'post']
+
+        if hasattr(cfg, 'norm_type'):
+            assert cfg.norm_type in ['mixedfusedlayernorm', 'layernorm']
+
+        for val in ['row_init_method', 'column_init_method']:
+            if hasattr(cfg, val):
+                assert cfg.get(val) in ['xavier', 'zero', 'normal']
+
+    def _state_dict(self, destination=None, prefix=None, keep_vars=False):
+        state_dict_ = {}
+
+        for name, module in self.enc_dec_model.named_modules():
+            if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                for adapter_key in self.adapter_name_keys:
+                    adapter_module = module.adapter_layer[adapter_key]
+                    state_adapter_key = ':'.join([name, adapter_key])
+                    state_dict_[state_adapter_key] = adapter_module.state_dict()
+        return state_dict_
+
+    def _load_adapters_weights(self, state_dict, strict=True):
+        if not self.loading_base_model:
+            # load adapters only
+            for name, module in self.enc_dec_model.named_modules():
+                if isinstance(module, adapter_mixins.AdapterModuleMixin):
+                    for adapter_key in self.adapter_name_keys:
+                        adapter_module = module.adapter_layer[adapter_key]
+                        state_adapter_key = ':'.join([name, adapter_key])
+                        # only load the adapters if they are in the state_dict
+                        if state_adapter_key in state_dict:
+                            adapter_module.load_state_dict(state_dict[state_adapter_key], strict)
+        else:
+            # load the base model once and change the loading to adapters only
+            super().load_state_dict(state_dict, strict=False)
+            self.loading_base_model = False
+
+    def setup_optimizer_param_groups(self):
```

Comment:
Did you test this with both megatron_amp_o2=True and False. `megatron_amp_o2=True` changes the optimizer a little bit.

Reply:
Not yet, I will test with both